### PR TITLE
StateMP accepts wires

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
In the new device interface, a device need not have wires. In the case when it does provide wires, it would be best if measurement processes that don't require wires at all (`state()`, `sample()`, `probs()`) know to expand over the extra wires. To be able to do this, they need to save the wires somewhere, and `self.wires` kinda makes sense for this:)

**Description of the Change:**
- Allow initialization of a `StateMP` with wires (now that it doesn't indicate a density-matrix measurement).
- When given a strict subset of wires, `process_state` pads the state with zeros for all other wire axes
- When given a different `wire_order` from `self.wires`, `process_state` will transpose the state to match the MP wires

**Note on wire ordering:**
I use the convention that `StateMP.wires` is the global wire order we'd like to return when used (so we can swap out measurement processes [here](https://github.com/PennyLaneAI/pennylane/blob/078c4503b87e988f0ca68344ff57687d36f3e7fa/pennylane/devices/experimental/default_qubit_2.py#L217)), and the `wire_order` given to `process_state` is the wire order of the `state` being passed in (which happens [here](https://github.com/PennyLaneAI/pennylane/blob/078c4503b87e988f0ca68344ff57687d36f3e7fa/pennylane/devices/qubit/simulate.py#L91), at a point where we do not know about the device or its wires).

**Changes for `pad` dispatching:**
1. I'm not sure what `qml.math.pad(x, like="torch")` was even calling, but the docs suggest to use `torch.nn.functional.pad` so I registered it
2. The tensorflow `pad` assumes padding over more than one dimension, so I "hacked" it to add an empty broadcast dimension if needed (in other words, turn `[pad]` into `[[pad]]`)

**Benefits:**
- `StateMP.process_state` is more powerful!
- Allows me to open a follow-up PR where DQ2 expands results for the above listed MPs

**Possible Drawbacks:**
- Doesn't support batching (yet, I can add in this PR if necessary, just wanted something that worked)

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -26,6 +26,11 @@
   process, `DensityMatrixMP`.
   [(#4558)](https://github.com/PennyLaneAI/pennylane/pull/4558)
 
+* The `StateMP` measurement now accepts a wire order (eg. a device wire order). The `process_state`
+  method will re-order the given state to go from the inputted wire-order to the process's wire-order.
+  If the process's wire-order contains extra wires, it will assume those are in the zero-state.
+  [(#)]()
```
Comment (timmysilv) [inline doc/releases/changelog-dev.md:32]:
```suggestion
  [(#4570)](https://github.com/PennyLaneAI/pennylane/pull/4570)
```

File: pennylane/measurements/state.py [hunk 2]
```diff
@@ -155,7 +156,29 @@ def shape(self, device, shots):
...
+        wires = self.wires
+        if not wires or wire_order == wires:
+            return state
+
+        if set(wire_order) == set(wires):
+            state = qml.math.reshape(state, (2,) * len(wire_order))
+            desired_axes = [wires.index(w) for w in wire_order]
+            return qml.math.flatten(qml.math.transpose(state, desired_axes))
+
+        if not wires.contains_wires(wire_order):
```
[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/measurements/state.py:163]: After a bit of offline discussion I'm okay with this :)

**Review by mudit2812 — APPROVED** [review]
Here's a rocket 🚀

File: pennylane/math/single_dispatch.py [hunk 3]
```diff
@@ -456,6 +456,15 @@ def _cond_tf(pred, true_fn, false_fn, args):
 )
 
 
+def _pad_tf(tensor, paddings, mode="CONSTANT", constant_values=0, name=None):
+    if ar.ndim(paddings) == 1:
+        paddings = (paddings,)
+    return _i("tf").pad(tensor, paddings, mode=mode, constant_values=constant_values, name=name)
```
Comment (rmoyard) [inline pennylane/math/single_dispatch.py:462]: I think it creates an error for one test:

```
FAILED tests/pauli/test_conversion.py::TestPhasedDecomposition::test_differentiability[matrices0-False] - TypeError: pad(): argument 'pad' (position 2) must be tuple of ints, but found element of type tuple at pos 0
```

**Review by rmoyard — COMMENTED** [review]
It looks good to me, but there is an issue with the TensorFlow padding. Happy to approve after it is fixed.

File: pennylane/measurements/state.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane/measurements/state.py:163]: updated the PR description to explain the choice.

**Review by rmoyard — COMMENTED** [review]
It looks good on my side, but can we double check that it does not break lightning for example?

**Comment by timmysilv** [comment]
just added batching. @rmoyard this shouldn't break lightning (or any plugin for that matter) because no old device uses `StateMP.process_state` to compute the state. They all use `QubitDevice.statistics` which [just returns device.state](https://github.com/PennyLaneAI/pennylane/blob/078c4503b87e988f0ca68344ff57687d36f3e7fa/pennylane/_qubit_device.py#L661-L662)

File: pennylane/math/single_dispatch.py (hunk 3 above)
Comment (timmysilv) [reply pennylane/math/single_dispatch.py:462]: fixed!

**Review by rmoyard — APPROVED** [review]
Thanks, great changes :100:

**Pull request merged**